src/client.rs

```diff
@@ -280,14 +293,26 @@ impl ClientFlow {
     pub fn set_idle_done(&mut self) -> Option<ClientFlowCommandHandle> {
         self.send_command_state.set_idle_done()
     }
+}
 
-    #[cfg(feature = "expose_stream")]
-    /// Return the underlying stream for debug purposes (or experiments).
-    ///
-    /// Note: Writing to or reading from the stream may introduce
-    /// conflicts with imap-flow.
-    pub fn stream_mut(&mut self) -> &mut AnyStream {
-        &mut self.stream
+enum ClientReceiveState {
+    Greeting(ReceiveState<GreetingCodec>),
+    Response(ReceiveState<ResponseCodec>),
+    // This state is set only temporarily during `ClientReceiveState::change_state`
+    Dummy,
+}
+
+impl ClientReceiveState {
+    fn change_state(&mut self) {
+        // NOTE: This function MUST NOT panic. Otherwise the dummy state will remain indefinitely.
+        let old_state = std::mem::replace(self, ClientReceiveState::Dummy);
+        let codec = ResponseCodec::default();
+        let new_state = Self::Response(match old_state {
+            Self::Greeting(state) => state.change_codec(codec),
+            Self::Response(state) => state,
+            Self::Dummy => unreachable!(),
+        });
+        *self = new_state;
```

Comment:
Could this take ownership of `self` to prevent having to include a `Dummy` variant?

Reply:
I don't think so. We can't move `ReceiveState` out of `ClientFlow`, so then we would need use something like `Option<ReceiveState>` and use `None` as temporary value.

With our current solution we have very good local reasoning that `Dummy` is an unreachable state.

Reply:
pity, it's unfortunate that this is necessary but I get it :( The only way I can think of to prevent having a Dummy variant is to use `unsafe { }`.

Reply:
This whole topic is a rabbithole, see https://github.com/rust-lang/rfcs/pull/1736. TLDR we can't have nice things because Rust supports unwinding. See also [this blog post](https://smallcultfollowing.com/babysteps/blog/2024/05/02/unwind-considered-harmful/) :p

However, I resolve this conversation. I already use the same `Dummy` trick for `ServerFlow` and got used to it.